src/EditorFeatures/Core/Intents/DeleteParameterIntentProvider.cs

```diff
@@ -0,0 +1,75 @@
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditorFeatures.Intents;
+
+[IntentProvider(WellKnownIntents.DeleteParameter, LanguageNames.CSharp), Shared]
+internal class DeleteParameterIntentProvider : IIntentProvider
+{
+    private readonly IGlobalOptionService _globalOptionService;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DeleteParameterIntentProvider(IGlobalOptionService globalOptionService)
+    {
+        _globalOptionService = globalOptionService;
+    }
+
+    public async Task<ImmutableArray<IntentProcessorResult>> ComputeIntentAsync(
+        Document priorDocument,
+        TextSpan priorSelection,
+        Document currentDocument,
+        IntentDataProvider intentDataProvider,
+        CancellationToken cancellationToken)
+    {
+        var changeSignatureService = priorDocument.GetRequiredLanguageService<AbstractChangeSignatureService>();
+        var contextResult = await changeSignatureService.GetChangeSignatureContextAsync(
+            priorDocument, priorSelection.Start, restrictToDeclarations: false, _globalOptionService.CreateProvider(), cancellationToken).ConfigureAwait(false);
+
+        if (contextResult is not ChangeSignatureAnalysisSucceededContext context)
+        {
+            return ImmutableArray<IntentProcessorResult>.Empty;
+        }
+
+        var parameterIndexToDelete = context.ParameterConfiguration.SelectedIndex;
+        var parameters = context.ParameterConfiguration.ToListOfParameters();
+        var isExtensionMethod = context.ParameterConfiguration.ThisParameter != null;
+
+        if (isExtensionMethod && parameterIndexToDelete == 0)
+        {
+            // We can't delete the 'this' parameter of an extension method.
+            return ImmutableArray<IntentProcessorResult>.Empty;
+        }
+
+        var newParameters = parameters.RemoveAt(parameterIndexToDelete);
+
+        var signatureChange = new SignatureChange(context.ParameterConfiguration, ParameterConfiguration.Create(newParameters, isExtensionMethod, selectedIndex: 0));
+        var changeSignatureOptionResult = new ChangeSignatureOptionsResult(signatureChange, previewChanges: false);
+
+        var changeSignatureResult = await changeSignatureService.ChangeSignatureWithContextAsync(context, changeSignatureOptionResult, cancellationToken).ConfigureAwait(false);
+        if (!changeSignatureResult.Succeeded)
+        {
+            return ImmutableArray<IntentProcessorResult>.Empty;
+        }
+
+        Contract.ThrowIfNull(changeSignatureResult.UpdatedSolution);
+        var changedDocuments = changeSignatureResult.UpdatedSolution.GetChangedDocuments(priorDocument.Project.Solution).ToImmutableArray();
+        return ImmutableArray.Create(new IntentProcessorResult(changeSignatureResult.UpdatedSolution, changedDocuments, EditorFeaturesResources.Change_Signature, WellKnownIntents.DeleteParameter));
```

Comment:
so what is the computation model for intents?  these can clearly be super expensive operations.  Is this work deferred until the user actually invokes teh intent?

Reply:
Currently they are invoked when they detect the intent.  I've discussed with them that these intents (this, rename, etc) are all expensive operations, but they're still in the prototype stage for everything beyond generate ctor.  I expect the intellicode side will have to modify the UX interaction before we roll these out fully.

Reply:
Sorry I must have missed the memo/announcement, please bear with my dump question: how does intents get exposed to users? I don't think I have seen them while dogfooding yet. Are they all enabled (at least for dogfooders)?